### PR TITLE
[runtime] refactor decoder, asr_model to support more platforms

### PR DESCRIPTION
**Pull request by robin1001**
refactor the decoder and asr model to support more platforms in the future, such as onnx,  j3, and so on.

TODO
- [x] undefined reference error fix.
- [x] asr model copy since there is inner state in asr_model now 
- [x] accuracy test, as the following, there is some bug in the old code when endpoint is triggerred. So the accuracy is slightlty different.
``` c++
  // 3. Cache feature for next chunk
  if (state == DecodeState::kEndBatch) {
    // TODO(Binbin Zhang): Only deal the case when
    // chunk_feats.size() > cached_feature_size_ here, and it's consistent
    // with our current model, refine it later if we have new model or
    // new requirements
    CHECK(chunk_feats.size() >= cached_feature_size);
    cached_feature_.resize(cached_feature_size);
    for (int i = 0; i < cached_feature_size; ++i) {
      cached_feature_[i] = std::move(
          chunk_feats[chunk_feats.size() - cached_feature_size + i]);
    }
  }
```

### CODE REVIEW

File: runtime/core/decoder/asr_decoder.h [hunk 1]
```diff
@@ -137,15 +132,7 @@ class TorchAsrDecoder {
...
   // cache feature
-  std::vector<std::vector<float>> cached_feature_;
   bool start_ = false;
-
-  torch::jit::IValue subsampling_cache_;
-  // transformer/conformer encoder layers output cache
-  torch::jit::IValue elayers_output_cache_;
-  torch::jit::IValue conformer_cnn_cache_;
-  std::vector<torch::Tensor> encoder_outs_;
-  int offset_ = 0;  // offset
```
Comment (xingchensong) [inline runtime/core/decoder/asr_decoder.h:148]: 把cache维护从decoder移到model是出于什么考量？

File: runtime/core/decoder/asr_model.cc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    // chunk_feats.size() > cached_feature_size here, and it's consistent
+    // with our current model, refine it later if we have new model or
+    // new requirements
+    cached_feature_.resize(cached_feature_size);
+    for (int i = 0; i < cached_feature_size; ++i) {
+      cached_feature_[i] =
+        chunk_feats[chunk_feats.size() - cached_feature_size + i];
+    }
+  }
+}
```
Comment (xingchensong) [inline runtime/core/decoder/asr_model.cc:45]: std::move(chunk_feats[...])?

File: runtime/core/decoder/asr_model.cc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+void AsrModel::ForwardEncoder(
+    const std::vector<std::vector<float>>& chunk_feats,
+    std::vector<std::vector<float>> *ctc_prob) {
+  ctc_prob->clear();
+  int num_frames = cached_feature_.size() + chunk_feats.size();
+  if (num_frames > right_context_ + 1) {
+    this->ForwardEncoderFunc(chunk_feats, ctc_prob);
+    this->CacheFeature(chunk_feats);
+  }
+}
```
Comment (xingchensong) [inline runtime/core/decoder/asr_model.cc:57]: 可能拆的更细一点更有利于代码复用，比拆成
```cpp
std::vector<torch::jit:IValue> input, output;
this->PrepareInput(chunk_feats, &input); // 对应 https://github.com/wenet-e2e/wenet/pull/993/files#diff-af8a37176d09d3fc9c34bd80ad43518791685f9edd0970b43833e7d4fdf5149aR88-R115
this->ForwardEncoderChunk(input, &output); // 对应 https://github.com/wenet-e2e/wenet/pull/993/files#diff-af8a37176d09d3fc9c34bd80ad43518791685f9edd0970b43833e7d4fdf5149aR117-R125
this->ForwardCtc(output, ctc_prob); // 对应 https://github.com/wenet-e2e/wenet/pull/993/files#diff-af8a37176d09d3fc9c34bd80ad43518791685f9edd0970b43833e7d4fdf5149aR127-R141
```

这样我们在写 j3_asr_model.cc 时只需要继承TorchAsrModel并override`ForwardEncoderChunk`,前面`PrepareInput`和后面`ForwardCtc`是一样的

Comment (xingchensong) [reply runtime/core/decoder/asr_model.cc:57]: 修改一下comment，拆成
```cpp
torch::Tensor feats, chunk_out;
this->PrepareFeats(chunk_feats, &feats); // 对应 https://github.com/wenet-e2e/wenet/pull/993/files#diff-af8a37176d09d3fc9c34bd80ad43518791685f9edd0970b43833e7d4fdf5149aR88-R108
this->ForwardEncoderChunk(feats, &chunk_out); // 对应 https://github.com/wenet-e2e/wenet/pull/993/files#diff-af8a37176d09d3fc9c34bd80ad43518791685f9edd0970b43833e7d4fdf5149aR109-R125
this->ForwardCtc(chunk_out, ctc_prob); // 对应 https://github.com/wenet-e2e/wenet/pull/993/files#diff-af8a37176d09d3fc9c34bd80ad43518791685f9edd0970b43833e7d4fdf5149aR127-R141
this->CacheFeature(chunk_feats);
```

File: runtime/core/decoder/asr_decoder.h (hunk 1 above)
Comment (robin1001) [reply runtime/core/decoder/asr_decoder.h:148]: cache type is related to model.

File: runtime/core/decoder/asr_model.cc (hunk 2 above)
Comment (robin1001) [reply runtime/core/decoder/asr_model.cc:45]: const type.

File: runtime/core/decoder/asr_model.cc (hunk 3 above)
Comment (robin1001) [reply runtime/core/decoder/asr_model.cc:57]: PrepareFeats is related to platform  or backend(onnx/j3)
ForwardEncoderChunk and ForwardCTC can be combined to one function.

File: runtime/core/utils/utils.cc [hunk 4]
```diff
@@ -29,4 +33,58 @@ float LogAdd(float x, float y) {
...
+      std::pair<T, int32_t>,
+      std::vector<std::pair<T, int32_t>>,
+      ValueComp<T>>
+      pq(ValueComp<T>(), std::move(heap_data));
+  for (int32_t i = k; i < n; ++i) {
+    if (pq.top().first < data[i]) {
+      pq.pop();
+      pq.emplace(data[i], i);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (robin1001) [reply runtime/core/utils/utils.cc:70]: 这样做避免了堆插入操作，复杂度更小一些。

File: runtime/core/decoder/torch_asr_model.cc [hunk 5]
```diff
@@ -14,32 +15,212 @@ namespace wenet {
...
-            << eos_ << " is bidirectional decoder "
-            << is_bidirectional_decoder_ << " num threads "
-            << num_threads;
+  LOG(INFO) << "Torch Model Info:";
+  LOG(INFO) << "\tsubsampling_rate " << subsampling_rate_;
+  LOG(INFO) << "\tright context " << right_context_;
+  LOG(INFO) << "\tsos " << sos_;
+  LOG(INFO) << "\teos " << eos_;
+  LOG(INFO) << "\tis bidirectional decoder " << is_bidirectional_decoder_;
+  LOG(INFO) << "\tnum threads " << num_threads;
```
[the remark this answers is not in the document]
Comment (robin1001) [reply runtime/core/decoder/torch_asr_model.cc:45]: Done

**Review by xingchensong — APPROVED** [review]

**Pull request merged**